[13.0] base_rest: New api

lmignon opened the pull request:
Foward port of #48

lmignon committed:
datamodel: New addon allowing you to define dataclass like models serialisable to/from json/yaml

lmignon committed:
base_rest: New API

lmignon committed:
datamodel: registry is always available a class level We no more need to get the registry from the context since it's always available on the schema class

lmignon committed:
datamodel: schema property of NestedModel must be a class This change fixes the incompatibility with the latest marshmallow version

lmignon committed:
datamodels: ensures that the Odoo environment is always available as a property of datamodel instances as well as on marshmallow schema instances.

lmignon committed:
datamodel: marshmallow 2.0 is released

lmignon committed:
[MIG] datamodel: Migrate from 12.0

HviorForgeFlow commented:
Not sure if related with this PR but at start running an instance I've got this error:
```
2020-05-26 11:24:37,780 1 CRITICAL test odoo.service.server: Failed to initialize database `test`. 
Traceback (most recent call last):
  File "/opt/odoo/custom/src/odoo/odoo/service/server.py", line 1179, in preload_registries
    registry = Registry.new(dbname, update_module=update_module)
  File "/opt/odoo/custom/src/odoo/odoo/modules/registry.py", line 86, in new
    odoo.modules.load_modules(registry._db, force_demo, status, update_module)
  File "/opt/odoo/custom/src/odoo/odoo/modules/loading.py", line 535, in load_modules
    model._register_hook()
  File "/opt/odoo/auto/addons/base_rest/models/rest_service_registration.py", line 62, in _register_hook
    self._build_controllers_routes(services_registry)
  File "/opt/odoo/auto/addons/base_rest/models/rest_service_registration.py", line 66, in _build_controllers_routes
    for service in self._get_services(controller_def["collection_name"]):
  File "/opt/odoo/auto/addons/base_rest/models/rest_service_registration.py", line 98, in _get_services
    return work.many_components()
  File "/opt/odoo/auto/addons/component/core.py", line 438, in many_components
    return [comp(work_context) for comp in component_classes]
  File "/opt/odoo/auto/addons/component/core.py", line 438, in <listcomp>
    return [comp(work_context) for comp in component_classes]
  File "/opt/odoo/auto/addons/connector_search_engine/components/mapper.py", line 20, in __init__
    exporter = work.index.exporter_id
AttributeError: 'WorkContext' object has no attribute 'index'
```

lmignon commented:
@HviorForgeFlow Thank you for the report. I'll take a look ASAP.This error is strange. The mapper component should not be returned by the lookup into the component registry since it's not registered for the rest service collection.. 
One I've found the issue I'll come back to you

lmignon committed:
[FIX] base_rest: lookup for service components registed for the given collection
This specific lookup is required to filter out component without collection. In the same time we must also filter out components that are not services

lmignon committed:
[FIX] base_rest: Avoid property resolution when inspecting object members
This is required to avoid some trouble with properties resolved by the method inspect.getmembers and accessing resources not yet completly initialized

lmignon commented:
@HviorForgeFlow The encountered issue should be resolved by the 2 last commits.

HviorForgeFlow commented:
It is working fine now thanks!

lmignon commented:
@HviorForgeFlow Thank you for the feedback. If you need to develop new services, I encourage you to use the new [api](https://github.com/OCA/rest-framework/pull/71/files#diff-446e66ad580b33ff05f11e44b939cd38R7) and [datamodels](https://github.com/OCA/rest-framework/pull/71/files#diff-c18d04b077d35612177ffcbec060125aR10). It's much more flexible and easier to maintain.

lmignon commented:
@guewen @gurneyalex The new API is now also available for Odoo 13. It would be nice if you could use-it / test-it with your new WMS mobile app :smirk:

kevinkhao committed:
Minor typo fixes

kevinkhao committed:
[IMP] Add decimal to float conversion for JSON serializing

kevinkhao committed:
[IMP] Return explanatory message for invalid JSON request instead of server error if a json request is made with bad json

gurneyalex commented:
hello @lmignon I see #48 is flagged as "in progress". Is it also the case for #71 ?

lmignon commented:
> hello @lmignon I see #48 is flagged as "in progress". Is it also the case for #71 ?
Good catch @gurneyalex. In fact, both of them needs review.

kevinkhao commented:
@lmignon before I do a PR, I'd like to know what you think.
Here is some context: https://github.com/acsone/rest-framework/blob/37b69406818a82deedccbb0444c731cf12a2fc20/base_rest/restapi.py#L194
It's about functions from_params and to_response.
For our controllers, when we use Datamodel and because we most probably just use my_input_datamodel.dump() the on the input and my_output_datamodel.load() before returning our result:
Would it make sense to:
1. dump our datamodel so that it is directly available in the controller's arguments ? i.e: in function `from_params`
`            return ModelClass.load(
                params, many=self._is_list, unknown=marshmallow.EXCLUDE
            ).dump()
`
2. allow to return a dict and it is automatically load()ed on the datamodel ? i.e in function `to_response`
remove lines 204 to 208
`        errors = ModelClass.validate(result, many=self._is_list)`
What do you think ?

Currently it does seem to me as bringing unnecessary steps, but I am maybe missing something.

lmignon commented:
@kevinkhao Sorry but I don't understand your question. The [`@Datamodel`](https://github.com/acsone/rest-framework/blob/37b69406818a82deedccbb0444c731cf12a2fc20/base_rest/restapi.py#L189) decorator is designed to decorate methods where the result of the call to the method is an instance of Datamodel and the param is also Datamodel instance. This decorator allows to only deal with datamodel objects on the service layer implementation.

nilshamerlinck committed:
[FIX] http/json error on first json request of new worker

kevinkhao commented:
> @kevinkhao Sorry but I don't understand your question. The [`@Datamodel`](https://github.com/acsone/rest-framework/blob/37b69406818a82deedccbb0444c731cf12a2fc20/base_rest/restapi.py#L189) decorator is designed to decorate methods where the result of the call to the method is an instance of Datamodel and the param is also Datamodel instance. This decorator allows to only deal with datamodel objects on the service layer implementation.

My idea was that when we define a route, we should just return a dict and the module will take care to do self.env.datamodel["my.output.datamodel"].load(resulting_dict) so that we can save 1 line at the end of our route definition. But after thinking about it, I'm still not sure about that idea.

Calling at the end of all my routes: 
return self.env["my.datamodel"].load(myResultDict)
instead of
return myResultDict
is more verbose but more clear.

lmignon commented:
> 
> My idea was that when we define a route, we should just return a dict and the module will take care to do self.env.datamodel["my.output.datamodel"].load(resulting_dict) so that we can save 1 line at the end of our route definition. But after thinking about it, I'm still not sure about that idea.
> 

@kevinkhao But we don't have to declare a route by ourself. Routes are dynamically generated by the framework based on the `@restapi.method` decorator https://github.com/OCA/rest-framework/pull/71/commits/262def2f7df2cd88653075f38964ae401d7dcbf7

lmignon committed:
Merge remote-tracking branch 'oca/13.0' into 13.0-48-port

lmignon committed:
[FIX] base_rest: fix external dependencies declaration

lmignon committed:
[FIX] base_rest: Use apispec>=4.0.0

HviorForgeFlow commented:
I've been testing new API spec you added in `base_rest_demo` module, I'm getting an error when trying to fetch that API collection in swagger:
```
2020-10-27 09:46:38,936 148419 INFO 13.0-rest-framework_new-api odoo.addons.base.models.ir_http: Generating routing map for key None 
2020-10-27 09:46:39,120 148419 INFO 13.0-rest-framework_new-api odoo.modules.registry: Registry changed, signaling through the database 
2020-10-27 09:46:39,122 148419 INFO 13.0-rest-framework_new-api werkzeug: 127.0.0.1 - - [27/Oct/2020 09:46:39] "GET /api-docs?urls.primaryName=base_rest_demo_api%2Fnew_api%3A%20partner HTTP/1.1" 200 - 184 0.041 0.147
2020-10-27 09:46:39,157 148419 INFO 13.0-rest-framework_new-api werkzeug: 127.0.0.1 - - [27/Oct/2020 09:46:39] "GET /web/content/430-c36b3b3/web.assets_common.js HTTP/1.1" 304 - 4 0.003 0.009
2020-10-27 09:46:39,160 148419 INFO 13.0-rest-framework_new-api werkzeug: 127.0.0.1 - - [27/Oct/2020 09:46:39] "GET /web/content/433-5375056/web.assets_frontend.js HTTP/1.1" 304 - 4 0.005 0.009
2020-10-27 09:46:39,167 148419 INFO 13.0-rest-framework_new-api werkzeug: 127.0.0.1 - - [27/Oct/2020 09:46:39] "GET /web/content/434-6e08f1c/base_rest.assets_swagger.js HTTP/1.1" 304 - 4 0.003 0.017
2020-10-27 09:46:39,238 148419 INFO 13.0-rest-framework_new-api werkzeug: 127.0.0.1 - - [27/Oct/2020 09:46:39] "GET /web/webclient/translations/1603791999227?mods=&lang=en_US HTTP/1.1" 200 - 5 0.003 0.005
2020-10-27 09:46:39,240 148419 INFO 13.0-rest-framework_new-api werkzeug: 127.0.0.1 - - [27/Oct/2020 09:46:39] "GET /web/webclient/locale/en_US HTTP/1.1" 200 - 1 0.000 0.005
2020-10-27 09:46:39,540 148419 INFO 13.0-rest-framework_new-api werkzeug: 127.0.0.1 - - [27/Oct/2020 09:46:39] "GET /api-docs/base_rest_demo_api/new_api/partner.json HTTP/1.1" 500 - 3 0.001 0.006
2020-10-27 09:46:39,545 148419 ERROR 13.0-rest-framework_new-api werkzeug: Error on request:
Traceback (most recent call last):
  File "/home/hvior/Projects/odoo/13.0/.env/lib/python3.6/site-packages/werkzeug/serving.py", line 270, in run_wsgi
    execute(self.server.app)
  File "/home/hvior/Projects/odoo/13.0/.env/lib/python3.6/site-packages/werkzeug/serving.py", line 258, in execute
    application_iter = app(environ, start_response)
  File "/home/hvior/Projects/odoo/13.0/odoo/odoo/service/server.py", line 439, in app
    return self.app(e, s)
  File "/home/hvior/Projects/odoo/13.0/odoo/odoo/service/wsgi_server.py", line 142, in application
    return application_unproxied(environ, start_response)
  File "/home/hvior/Projects/odoo/13.0/odoo/odoo/service/wsgi_server.py", line 117, in application_unproxied
    result = odoo.http.root(environ, start_response)
  File "/home/hvior/Projects/odoo/13.0/odoo/odoo/http.py", line 1287, in __call__
    return self.dispatch(environ, start_response)
  File "/home/hvior/Projects/odoo/13.0/odoo/odoo/http.py", line 1257, in __call__
    return self.app(environ, start_wrapped)
  File "/home/hvior/Projects/odoo/13.0/.env/lib/python3.6/site-packages/werkzeug/wsgi.py", line 766, in __call__
    return self.app(environ, start_response)
  File "/home/hvior/Projects/odoo/13.0/odoo/odoo/http.py", line 1457, in dispatch
    result = ir_http._dispatch()
  File "/home/hvior/Projects/odoo/13.0/odoo/addons/web_editor/models/ir_http.py", line 21, in _dispatch
    return super(IrHttp, cls)._dispatch()
  File "/home/hvior/Projects/odoo/13.0/odoo/odoo/addons/base/models/ir_http.py", line 238, in _dispatch
    return cls._handle_exception(e)
  File "/home/hvior/Projects/odoo/13.0/odoo/odoo/addons/base/models/ir_http.py", line 206, in _handle_exception
    return request._handle_exception(exception)
  File "/home/hvior/Projects/odoo/13.0/odoo/odoo/http.py", line 750, in _handle_exception
    return super(HttpRequest, self)._handle_exception(exception)
  File "/home/hvior/Projects/odoo/13.0/odoo/odoo/http.py", line 310, in _handle_exception
    raise pycompat.reraise(type(exception), exception, sys.exc_info()[2])
  File "/home/hvior/Projects/odoo/13.0/odoo/odoo/tools/pycompat.py", line 14, in reraise
    raise value
  File "/home/hvior/Projects/odoo/13.0/odoo/odoo/addons/base/models/ir_http.py", line 234, in _dispatch
    result = request.dispatch()
  File "/home/hvior/Projects/odoo/13.0/odoo/odoo/http.py", line 809, in dispatch
    r = self._call_function(**self.params)
  File "/home/hvior/Projects/odoo/13.0/odoo/odoo/http.py", line 350, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/home/hvior/Projects/odoo/13.0/odoo/odoo/service/model.py", line 94, in wrapper
    return f(dbname, *args, **kwargs)
  File "/home/hvior/Projects/odoo/13.0/odoo/odoo/http.py", line 339, in checked_call
    result = self.endpoint(*a, **kw)
  File "/home/hvior/Projects/odoo/13.0/odoo/odoo/http.py", line 915, in __call__
    return self.method(*args, **kw)
  File "/home/hvior/Projects/odoo/13.0/odoo/odoo/http.py", line 515, in response_wrap
    response = f(*args, **kw)
  File "/home/hvior/Projects/odoo/13.0/addons/OCA/rest-framework/base_rest/controllers/api_docs.py", line 38, in api
    return self.make_json_response(service.to_openapi())
  File "/home/hvior/Projects/odoo/13.0/addons/OCA/rest-framework/base_rest/components/service.py", line 183, in to_openapi
    return BaseRestServiceAPISpec(self).to_dict()
  File "/home/hvior/Projects/odoo/13.0/addons/OCA/rest-framework/base_rest/apispec/base_rest_service_apispec.py", line 36, in __init__
    self._generate_paths()
  File "/home/hvior/Projects/odoo/13.0/addons/OCA/rest-framework/base_rest/apispec/base_rest_service_apispec.py", line 74, in _generate_paths
    self._add_method_path(method)
  File "/home/hvior/Projects/odoo/13.0/addons/OCA/rest-framework/base_rest/apispec/base_rest_service_apispec.py", line 66, in _add_method_path
    routing=routing,
  File "/home/hvior/Projects/odoo/13.0/.env/lib/python3.6/site-packages/apispec/core.py", line 280, in path
    plugin.operation_helper(path=path, operations=operations, **kwargs)
  File "/home/hvior/Projects/odoo/13.0/addons/OCA/rest-framework/base_rest/apispec/rest_method_param_plugin.py", line 59, in operation_helper
    responses.update(output_param.to_openapi_responses(self._service))
  File "/home/hvior/Projects/odoo/13.0/addons/OCA/rest-framework/base_rest/restapi.py", line 231, in to_openapi_responses
    "application/json": {"schema": self.to_json_schema(service)}
  File "/home/hvior/Projects/odoo/13.0/addons/OCA/rest-framework/base_rest/restapi.py", line 238, in to_json_schema
    schema = self._get_schema(service)
  File "/home/hvior/Projects/odoo/13.0/addons/OCA/rest-framework/base_rest/restapi.py", line 242, in _get_schema
    return service.env.datamodels[self._name].get_schema(many=self._is_list)
  File "/home/hvior/Projects/odoo/13.0/addons/OCA/rest-framework/datamodel/core.py", line 383, in __getitem__
    model = self.registry[key]
TypeError: 'NoneType' object is not subscriptable - - -
```
![image](https://user-images.githubusercontent.com/32061121/97285310-551ffd00-1842-11eb-97a8-d3809053aa88.png)

lmignon commented:
> I've been testing new API spec you added in `base_rest_demo` module, I'm getting an error when trying to fetch that API collection in swagger:
> 
> ```
> 2020-10-27 09:46:38,936 148419 INFO 13.0-rest-framework_new-api odoo.addons.base.models.ir_http: Generating routing >     return service.env.datamodels[self._name].get_schema(many=self._is_list)
>   File "/home/hvior/Projects/odoo/13.0/addons/OCA/rest-framework/datamodel/core.py", line 383, in __getitem__
>     model = self.registry[key]
> TypeError: 'NoneType' object is not subscriptable - - -
> ```
> 
> ![image](https://user-images.githubusercontent.com/32061121/97285310-551ffd00-1842-11eb-97a8-d3809053aa88.png)

It seems that the datamodel registry is not loaded. Do you have more than one odoo13 database available into your postgresql db? To work properly, odoo should only get access to one and only one database. That could be the origin of this issue.

It works as expected on runbot.... http://3446717-71-7cfcaf.runbot3.odoo-community.org/api-docs#/

HviorForgeFlow commented:
OK, thanks for the advice, I couldn't figure it out.

etobella committed:
[FIX] base_rest: finish new rest API

simahawk commented on base_rest/components/service.py:
```diff
@@ -91,114 +80,86 @@ def _log_call(self, func, params, secure_params, res):
                        to the method as keyword args.
         :return:
         """
+        method = getattr(self, method_name, object())
```
why `object` as default?

simahawk commented on base_rest/components/service.py:
```diff
@@ -91,114 +80,86 @@ def _log_call(self, func, params, secure_params, res):
-        return validator.schema
-
-    def _secure_input(self, method, params):
+    def _prepare_input_params(self, method, params):
```
https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings

for all the docstrings

simahawk commented on base_rest/models/rest_service_registration.py:
```diff
@@ -48,6 +57,52 @@ def _register_hook(self):
+
+        # generate an addon name used to register our new controller for
+        # the current database
+        addon_name = "{}_{}".format(
```
shouldn't belong to the service's original module?

lmignon replied:
Since services can be inherited and to avoid having to search in which addon the service was first declared, it is easier to build this name on the basis of the service's information. The goal is to have 1 controller per service. Reading the code again, it makes me say that the name of the collection should also be included in the name of the addon because the same service can be declared in several collections....

simahawk commented on base_rest/models/rest_service_registration.py:
```diff
@@ -81,3 +136,212 @@ def _init_global_registry(self):
+
+
+class RestApiMethodTransformer(object):
+    """ Helper class to generate and apply the missing retsapi.method decorator
```
```suggestion
    """ Helper class to generate and apply the missing restapi.method decorator
```

simahawk commented on base_rest/models/rest_service_registration.py:
```diff
@@ -81,3 +136,212 @@ def _init_global_registry(self):
+        """
+        return type(self._new_cls_name, (self._base_controller,), self._get_methods())
+
+    def _get_methods(self):
```
:dizzy_face:  :sweat_smile:

simahawk commented on base_rest/restapi.py:
```diff
@@ -0,0 +1,250 @@
+        """
+
+        :param schema: can be dict as cerberus schema, an instance of
+                       cerberus.Validator or a sting with the method name to
```
```suggestion
                       cerberus.Validator or a string with the method name to
```

simahawk commented on base_rest/restapi.py:
```diff
@@ -0,0 +1,250 @@
+        :param schema: can be dict as cerberus schema, an instance of
+                       cerberus.Validator or a sting with the method name to
+                       call on the service to get the schema or the validator
+        :param is_list:
```
used for?

simahawk commented on datamodel/core.py:
```diff
@@ -0,0 +1,401 @@
+        ::
+
+            from marshmallow import fields
+            from odoo.addons.datamode.core import Datamodel
```
```suggestion
            from odoo.addons.datamodel.core import Datamodel
```

simahawk commented on datamodel/core.py:
```diff
@@ -0,0 +1,401 @@
+
+@property
+def datamodels(self):
+    class DataModelFactory(object):
```
why is this class inside the property?

simahawk commented on datamodel/tests/common.py:
```diff
@@ -0,0 +1,224 @@
+    need to explicitly pass ``self.datamodel_registry`` in the
+    """
+
+    def setUp(self):
```
I've reworked these classes here to allow components registration on setupclass instead of setUp only https://github.com/OCA/connector/pull/387

BTW SavepointCase is going to disappear in master and TransactionCase will use setUpClass

simahawk commented on base_rest/restapi.py:
```diff
@@ -0,0 +1,250 @@
+from .tools import cerberus_to_json
+
+
+def method(
```
:heart:

simahawk reviewed (commented):
@lmignon amazing (and crazy) work :smile: :heart: 
I have just few remarks.
Any real life usage so far beside https://github.com/shopinvader/odoo-shopinvader/pull/802 ?

lmignon commented:
Thank you @simahawk for the review. Unfortunately I can't find time to finalise all this work since it's only done on my spare time and I'm overloaded... There are others comments that I've to also take into account from #48 . I hope to be able to find some times in the future to finish this work but when????

kevinkhao commented:
@simahawk if you're looking for another example that uses it
https://github.com/akretion/sale-import/tree/12.0/sale_import_rest

lmignon commented:
@kevinkhao @sebastienbeau @etobella @simahawk Before merging all these new stuff, I wonder if it would be better to separate the 'datamodels' part in a dedicated addon and only support cerberus into base_rest. I have the feeling that this will facilitate the maintenance of these modules by bringing a better modularity and having a lightweight base addon. It will also prevent those who don't want to use datamodels from having to install a set of dependencies they don't need.

simahawk commented:
@lmignon I was about to ask you this and split the datamodels part to another PR as well. It would speed up things and ease review and maintenance from other people.

sbidoul commented:
> I wonder if it would be better to separate the 'datamodels' part in a dedicated addon and only support cerberus into base_rest

I probably don't get all the implications but I'm wondering: is it not the plan that the datamodel api would replace cerberus in the long term ? If that is the intent, it is probably better to have it very visible in base_rest so people adopt it.

lmignon committed:
[IMP] base_rest: Move code to support datamodel into a dedicated addon

lmignon committed:
[ADD] setup.py

lmignon committed:
[FIX] pre-commit: Fix pylint

lmignon committed:
[IMP] base_rest: Ensure controller unicity by service and collection

lmignon committed:
[IMP] base_rest, datamode: Spelling

lmignon committed:
[IMP] base_rest: Improve code readability

lmignon committed:
[IMP] datamodel: Move DataModelFactory class definition outside the datamodels method implementation

lmignon committed:
[FIX] base_rest_datamodel: Upgrade to 13.0

lmignon committed:
[IMP] base_rest: Improve documentation

lmignon commented:
@simahawk @sebastienbeau I worked on this PR again so that we could merge it.  The datamodel part is now in a separate addon. I have added some documentation and fixed some small problems. I would like to merge this PR in 13.0 and have it as the basis for 14.0.

lmignon committed:
[FIX] base_rest: Add missing http routes for service methods declared without the restapi.method decorator

lmignon committed:
[IMP] base_rest: Add tests for the openapi document generation process

lmignon committed:
[IMP] base_rest: Add tests for CerberusValidator

lmignon committed:
[FIX] base_rest: Adapt tests according to changes into component

lmignon committed:
fix website into __manifest__.py

lmignon committed:
[IMP] base_rest: remove deprecated call to inspect.getargspec

simahawk commented:
@lmignon sorry, was on holidays and missed your msg. I'll skim through your last changes in a while.

simahawk commented on base_rest/tests/common.py:
```diff
@@ -32,6 +42,67 @@ def setUpRegistry(cls):
             )
 
 
+class RestServiceRegistryCase(ComponentRegistryCase):
```
You can use Savepoint case as well now :wink:

simahawk commented on base_rest/tests/common.py:
```diff
@@ -49,56 +50,71 @@ def setUpRegistry(cls):
+    # pylint: disable=W8106
+    @staticmethod
+    def _setup_registry(class_or_instance):
+        super().setUpClass()
```
setUpClass call here? It should be the other way around: setUpClass calls _setup_registry, no? :thinking:

lmignon replied:
Good catch @simahawk I removed the line since it's useless and is a remaining of the refactoring.

lmignon committed:
[FIX] base_rest: Remove wrong call to setUpClass into RestServiceRegistryCase

lmignon commented:
@hparfr, @sebastienbeau Are you OK with this PR?

hparfr reviewed (approved):
The tests I had conducted so far on WMS shows that it seams to work.
I will do more testing when porting shopinvader soon and open issues or propose PR if needed.

simahawk commented:
@lmignon good to merge then?

lmignon commented:
/ocabot merge major

sbidoul commented:
![fireworks](https://user-images.githubusercontent.com/692075/104740068-c8e9eb80-5747-11eb-89d9-4ef8928af851.gif)

sebastienbeau commented:
youhou \o/
